source/MRMesh/MRObjectLoad.cpp

```diff
@@ -96,79 +100,179 @@ tl::expected<ObjectDistanceMap, std::string> makeObjectDistanceMapFromFile( cons
     return objectDistanceMap;
 }
 
-tl::expected<Object, std::string> makeObjectTreeFromFolder( const std::filesystem::path & folder )
+bool isAnySupportedFilesInSubfolders( const std::filesystem::path& folder )
+{
+    std::vector<std::filesystem::path> filesList;
+    filesList.push_back( folder );
+
+    while ( !filesList.empty() )
+    {
+        auto path = filesList[0];
+        filesList.erase( filesList.begin() );
+
+        std::error_code ec;
+        for ( auto& directoryEntry : std::filesystem::directory_iterator( path, ec ) )
+        {
+            auto subpath = directoryEntry.path();
+            if ( directoryEntry.is_directory( ec ) )
+            {
+                filesList.push_back( path = subpath );
+            }
+            else if ( directoryEntry.is_regular_file( ec ) )
+            {
+                auto ext = utf8string( subpath.extension() );
+                for ( auto& c : ext )
+                    c = ( char )tolower( c );
+
+                if ( ext.empty() )
+                    continue;
+
+                if ( std::find_if( allFilters.begin(), allFilters.end(), [&ext] ( const IOFilter& f )
+                    { return f.extension.substr( 1 ) == ext; } ) != allFilters.end() )
+                    return true;
+            }
+        }
+    }
+    return false;
+}
+
+tl::expected<Object, std::string> makeObjectTreeFromFolder( const std::filesystem::path & folder, ProgressCallback callback )
 {
     MR_TIMER;
 
-    Object root;
+    if ( callback && !callback( 0.f ) )
+        return getCancelMessage( folder );
 
-    struct LoadTask
+    struct FilePathNode
     {
-        std::future< tl::expected<ObjectMesh, std::string> > future;
-        Object * parent = nullptr;
-        LoadTask( std::future< tl::expected<ObjectMesh, std::string> > future, Object * parent ) : future( std::move( future ) ), parent( parent ) { }
+        std::filesystem::path path;
+        std::vector<FilePathNode> subfolders;
+        std::vector<FilePathNode> files;
     };
-    std::vector<LoadTask> loadTasks;
 
-    struct Subfolder
+    FilePathNode filesTree;
+    filesTree.path = folder;
+
+
+    std::function<void( FilePathNode& )> fillFilesTree = {};
+    fillFilesTree = [&fillFilesTree] ( FilePathNode& node )
+	{
+		std::error_code ec;
+        for ( auto& directoryEntry : std::filesystem::directory_iterator( node.path, ec ) )
```

Comment:
this is unsafe, `operator ++` for directory iterators throws exception, use `increment( , ec)` as in other places

Reply:
fixed